### DISCUSSION TITLE
Не все эффекты работают на версии 1.14

### DISCUSSION THREAD

**Discussion by smolin1964** [vvip-68/LedPanelWiFi#57] [Вопросы и ответы] [answered]
Доброе время суток!!! Установил версию 1.14.  Гирлянда расположена на двух окнах. Соединил по протоколу E131. Одну сделал мастером, другую ведомой. Тип синхронизации локальный ( на версии 1.13 делал также). Все гирлянды засинхронизировались на ведомой надпись "слава". Обратил внимание что на мастере на работают несколько эффектов. Если включаю автономный (на мастере) все работает. При этом на ведомом эффекты работают, а на мастере остается прежний эффект (замерзший).  Кто может подсказать в чем проблема???? Не работают эффекты-"Облака, Лава, Плазма, Бензин на воде,Павлин, Зебра, Шумный лес, Морской прибой".  Заранее спасибо!!

**Comment by vvip-68** [comment]
Проблема известная, наблюдалась и у меня тоже. Причина - неизвестна.

Но у меня, правда, наблюдалась только в основном в автономном режиме и то может быть только потому, что у меня на окнах три Slave, a Master вообще ни к чему не подключен, может поэтому не видел. 

В автономном режиме у меня выглядело так - при переключении на некоторые эффекты на матрице остается предыдущий и только мерцает первый светодиод в цепочке. Включаешь бегущую строку - начинает показываться. По завершению бегущей строки снова только мерцание первого светодиода.

А бывает и наоборот - все показывается, но когда начинается бегучая строка - эффект замирает, только мерцает первый светодиод.

Повторюсь - причину я не знаю. Методы лечения - примерно следующие:
- в произвольном месте скетча вставить 5-10-15 пустых строк
- в произвольном месте скетча вставить ничего не делающие операторы типа объявления переменной string tmp; tmp.reserve(20);
- поперемещать (попереставлять местами) блоки (функции) в скетче.
- В меню "Инструменты" поперебирать параметры компиляции в зависимости от выбранной платы, например для ESP8266 - что-то вроде "Включить контроль стека" или "Контроль выделения памяти OOM" Enabled / Disabled. Можно поперебирать Flash Mode DOUT..QOUT. Поизменять частоту процессора / флэш-памяти.

При какой-нибудь комбинации обычно ситуация выправляется.
Повторюсь - причина мне неизвестна. Отсюда - вместо осознанных методов борьбы - вот такие магические и нелогичные действия, которые, впрочем, в большинстве случаев работают.

**Comment by vortigont** [comment]
причина - некорректная работа с памятью, отсутствие блокировок и висячие указатели. Где-то вы переаллоцируете память, а указатели продолжают смотреть в старое место.

**Reply by vvip-68** [reply]
Мониторинг памяти утечек не показывает.

До "оптимизации" и применения оптимизированной AsyncWebServer из вашего репозитория этого не замечалось.
Найти ошибки внутри библиотек мне не под силу.

Почему такое влияние - только на FastLED, который не может отправить сигнал далее первого  светодиода?
Ему по идее должно быть все равно какой "мусор" выводить, если память не там аллоцирована. О том, что эффект формирует картинку правильно и проблема только в FastLED - сформированное изображение эффекта успешно отправляется по сети на ведомое устройство и отображается им правильно.
Поведение FastLED в данном случае точно такое же, как после выхода новой версии ядер 3.x.x для 8266 и 2.0.0 для ESP32, когда FastLED вел себя точно так же - не было вывода на матрицу, проблемы с первым светодиодом.

Почему, не трогая кода никакого перераспределения памяти, просто добавление пустых строк или пустых операторов выправляет ситуацию? Почему, если указатели указывают не туда, это не влияет ни на что кроме FastLED и сама программа при этом не падает? Почему это касается не какого-то определенного эффекта, а может случиться на любом? 

Память под эффект выделяется в начале его отображения и освобождается после. В ходе "воспроизведения перераспределения памяти нет". Почему вывод эффекта возобновляется, когда начинается отображение бегущей строки - она вообще никакой памяти не аллоцирует.

Про блокировки - это наблюдается и на одноядерном 8266, где нет никаких параллельных и конкурирующих процессов. При чем тут блокировки?

**Comment by vortigont** [comment]
> Мониторинг памяти утечек не показывает.

утечек может и не быть, "висячий" указатель ничем не отличается от валидного если архитектура процессора не защищает сегметы памяти разных процессов или если указатель остается в рамках своего сегмента.

> До "оптимизации" и применения оптимизированной AsyncWebServer из вашего репозитория этого не замечалось.

асинк-сервер никакого отношения непосредственно к фастед не имеет, не надо валить с больной головы на здоровую. В асинке есть свои проблемы, но "портить" фастлед он не умеет. Проблема в том что мало кто отдает себе отчет при использовании в его асинхронности. Т.е. если где-то в его колбеках ВЫ в _своем_ коде делаете (ре)аллокацию памяти, вы потенциально имеете проблемы испортить свои же указатели.

> Почему такое влияние - только на FastLED, который не может отправить сигнал далее первого светодиода?

моргающий первый диод означает лишь то что фастлед выводит на ленту какой-то блок памяти в котором меняется только хвост в несколько байт. Т.е. грубо говоря, эффект ваш пишет новые данные в одно место в памяти, а фастлед выводит данные их другого.
Фастлед не знает выводит он данные из аллоцированного участка или нет, у него есть указатель на начало сегмента памяти и он шпарит оттуда байты определенной длинны. Архитектура есп допускает такие вольности, на других архитектурах, вероятно, код упал бы с сегментейшен фолт.

> Почему, не трогая кода никакого перераспределения памяти, просто добавление пустых строк или пустых операторов выправляет ситуацию? 

по-разному слинкованый код смещает адреса расположения объектов в памяти. Когда повезет указатель расположен там где надо, когда не повезет - не там :) такое поведение как раз и указывает на то что проблемы где-то с использованием неверных указателей.

> Почему, если указатели указывают не туда, это не влияет ни на что кроме FastLED и сама программа при этом не падает?

вероятно потому что вы в основном выделяете/меняете память под то что связанно с фастлед - память под буфера с данными для светодиодов. Это сырые данные тупо на чтение. В остальном вы, в основном, пользуетесь библиотечными вызовами. Хотя на это можно наступить и в других местах - с ардуиноджейсон например. Но выражаться может по-другому - или мусор в данных или падение если используются указатели на объекты.

> Почему это касается не какого-то определенного эффекта, а может случиться на любом?

именно поэтому. Вы переключаете эффекты и (ре)аллоцируете какие-то буфера


> это наблюдается и на одноядерном 8266, где нет никаких параллельных и конкурирующих процессов. При чем тут блокировки?

асинк-сервер он и на 8266 **асинк**, т.е. его колбеки могут быть вызваны прямо "посередине" обсчета эффекта, например, как повезёт. Он работает в сетевом стеке, WiFi работает "параллельно" коду в loop() и когда надо прерывает его работу.
Т.е. **к примеру** у вас есть какой-то цикл, который считает "красоту" и пишет данные в какой-то 3Д шум буфер какого-то эффекта. Где-то в середине цикла прилетает запрос на переключение эффекта, в обработчике запроса вы, _к примеру_, решаете переключить эффект и освободить буфер под 3Д шум и аллоцировать какой-то другой буфер. После обработчика сервера ваш цикл эффекта продолжает работу обсчета и пишет данные по _старому_ указателю. Может сделать еще что-то, может переписать какой-то байт в памяти куда во время обработки вызова асинк-сервера был записан указатель на адрес _нового_ буфера под фастлед. Предстказать что может произойти довольно сложно, поэтому проблемы плавающие.

**Reply by vvip-68** [reply]
Еще раз. Эффект формируется правильно. FastLED отображает (пытается его отобразить) сразу после того как он отправлен в сеть и правильно принят ведомым устройством. Отображается и отправляется - один и тот же буфер. Поскольку мигает только первый светодиод - то хвост этот - ровно три байта. Если бы было больше - мигало бы несколько светодиодов. Да и вообще фастледу пофигу на выделение блока - он бы просто выводил мусор за границами выделенного блока, указанное количество байт но этого не происходит.

Как-то это всё звучит очень неубедительно.

**Comment by vortigont** [comment]
> Отображается и отправляется - один и тот же буфер.

вот тут вы ошибаетесь. В сеть вы отправляете данные _другим_ кодом из буфера в памяти по указателю. Фастлед к отправке данных в сеть _никакого_ отношения не имеет. Он выводит данные на ленту по _своему_ указателю. Именно поэтому у вас на слейвах всё бегает, а на мастере статика - фастлед выводит данные из блока в памяти в котором ничего не меняется при обсчете эффекта. Хвост в 3 байта это один выровненый 32-битный дворд. 



> Да и вообще фастледу пофигу на выделение блока - он бы просто выводил мусор за границами выделенного блока

мусор туда должен кто-то записать. Если блок памяти был "освобожден", то его возвращают в кучу, это не значит что туда тут же попадет мусор. Поэтому у вас на матрице остается статичная картина. Можете попробовать "заливать"  все свои цветовые буфера каким-нибудь цветом перед освобождением, будет видно наглядно.

> Как-то это всё звучит очень неубедительно.

ну да, зато магия с добавлением пустых строк как метод борьбы с глюкавым асинком намного более убедительнее :)
Я пояснил откуда берутся такие проблемы - дальше думайте сами.

**Reply by vvip-68** [reply]
При старте прошивки под буфер светодиодов выделяется память. Это leds[], указатель на который передается FastLED и из которого он выводит данные на светодиоды. Это один и тот же указатель на протяжении всей работы приложения и не меняется, не реаллоцируются.

Эффекты свою красоту пишут именно в этот буфер. В сеть отправляются данные, копируемые из ЭТОГО буфера, в буфер отправляющего. Причем на приемнике эти скопированные данные отображаются правильно, а вот сам фастлед уже почему-то не может отобразить этот буфер. Я бы понял, если на приемнике было что-то не то - не туда скопировали для отправки, не  тот указатель отправили.  Что-то не срастается...

**Comment by vortigont** [comment]
> Это один и тот же указатель на протяжении всей работы приложения и не меняется, не реаллоцируются.

Вот же у вас прямо в луп торчит аллокатор этого указателя который "не меняется" с припиской про "временное" освобождение
https://github.com/vortigont/LedPanelWiFi/blob/main/firmware/LedPanelWiFi_v1.14/LedPanelWiFi_v1.14.ino#L707-L714

детально я весь код не смотрел, но проблемных мест, как мне кажется, навалом. Если не хотите чтобы этот массив менялся аллоцируйте его статически и указатель объявляйте как конст.

**Reply by vvip-68** [reply]
Он выделяется один раз. Его освобождение происходит в критических случаях нехватки памяти и сопровождается выводом сообщения об этом в лог. Таких сообщений нет. Памяти (особенно на ESP32) - достаточно. Это было нужно для ESP8266, когда до оптимизаций свободными оставалось 5-7 килобайт. Сейчас постоянно свободно порядка 20K. Так что в этот кусок кода не выполняется никогда, поскольку выделение памяти единожды в setup() при старте прошивки и далее никогда память не освобождается и не выделяется. Теоретически на текущий момент его вообще можно выкинуть.

**Reply by vortigont** [reply]
я не сказал что именно _эта_ проблема именно в _этой_ строке, я указал на место в коде в ответ на ваше же утверждение что этот указатель никогда не переписывается. А насколько это вероятно или не вероятно спорить глупо - код есть, он может сработать. Не нужен - уберите, нужен - убедитесь что это безопасно.
С вами оч сложно вести диалог, вы скорее поверите в динозавров и рептилоидов чем в то что у вас в коде что-то не так. Я объяснил _механизм_ возникновения _подобных_ проблем, даже привел алгоритм как именно это просиходит. Согласны вы с этим или нет никак на сами причины и принцип работы асинк-сервера не влияет. У вас в коде полно подобных мест. У вас вся сцена увешана заряженными ружьями, в потолке дыра, а вы рассуждаете о том что никому стрелять из них не надо, это рептилоды проделали дыру в потолке. Это всё 100 раз вылезало во всех "лампах" здесь которые копипастятся одна с другой, все ноют на последний "мигающий диод" и замёрзшие кадры.
В самом первом посте товарищ привет список эффектов, насколько я помню, почти во всех из них используются дополнительные буфера, которые вы аллоцируете/освобождаете - это повод задуматься. Проблема возникает при "смене" эффектов - повод задуматься.
Вы же сами сказали что когда картинка эффекта "замерзает" то бегущая строка у вас _работает_, т.е. библиотека фастлед _способна_ выводить данные корректно. Где у вас теряются данные кадра самого эффекта - в оверлее, в промежуточных буферах эффектов, при копировании из одного в другое или еще где надо сидеть и разбираться в _своём_ коде. Или не разбираясь в бардаке переписывать блоками проблемные места так как положенно. Ну или забить и списать на "кривые" библиотеки и продолжать втыкать пустые строки и брызгать святую воду :)
Я у себя воспроизводил этот эффект, воспроизводил и более сложные случаи когда контроллер падал при деструкте/инстациации классов эффектов. У вас классов нет, поэтому у вас "картинка" замирает. Исправляется это всё или флагами или блокировками или и тем и другим.

**Reply by vvip-68** [reply]
1. Убрал. Посмотрим поменяется ли что-то. Хотя это сложно, потому что это сродни добавлению/убиранию пустых строк или передвижения кода вверх/вниз относительно друг друга. Может ничего и не показать, если эффект не проявится. Если проявится - тогда дело точно не в этом куске кода. И я на 100% уверен, что не в этом дело, так как этот код не вызывается.

2.  Я охотно верю, что в коде может быть что-то не так. Просто есть вероятность, что не в моем коде, а в коде ядра или библиотек. А такое предположение дает пройденный факт, что ранее - не меняя свой код, а заменив библиотеки или версии ядра - проблема уходила или появлялась. То что стабильно годами работало на 2.7.4 для 8266 или на 1.0.6 для ESP32 вдруг резко перестало работать вообще при смене ядра на 3.0.0 / 2.0.0 И после так же необъяснимым образом как-то стало работать на 3.2.1 / 2.0.14. С различными изменениями в работоспособности при замене ядер, скажем 2.0.2 -> 2.0.6 -> 2.0.14, не меняя кода приложения.

3. Я бы поверил, что в проблема в буфере фастледа, если бы на удаленном контроллере данные из этого самого буфера не отображались бы корректно.  А мигающий первый диод про который вы утверждаете что проблема в буфере - на ранних версиях ядер проявлялся и когда буфер выделялся статически и в коде не было ничего кроме - записать в буфер из/для трех диодов цвет и вывести на ленту (примеры из библиотеки FastLED).

4.  Список эффектов у товарища - один. У меня может быть и другой. У меня часть однотипных эффектов - все эти шумовые) - часть из них могла работать, часть вот  так вот мерцать первым диодом, хотя там используется один и тот же код, просто с разными палитрами.

**Reply by vvip-68** [reply]
Можете найти ошибку в моем коде, приводящую к указанному эффекту - найдите и ткните носом, тем более, что у себя вы ее воспроизводили, знаете что и где может пойти не так. Не можете (нет времени, желания - понимаю, чего в чужом коде ковыряться) - оставим как есть. Я причины найти не могу. Буду двигать строки кода и брызгать святой водой. Топикстартеру могу только порекомендовать оставить мою прошивку - я ее никому не навязываю - и перейти на идеально работающую и безглючную Вашу.

**Comment by vortigont** [comment]
> 2\. не меняя свой код, а заменив библиотеки или версии ядра - проблема уходила или появлялась

ну так это тоже самое что повтыкать пробелов в код. Проблема не ушла, проблема просто пока не проявилась.

> 3\. на ранних версиях ядер проявлялся и когда буфер выделялся статически и в коде не было ничего кроме - записать в буфер из/для трех диодов цвет и вывести на ленту (примеры из библиотеки FastLED)

давайте рассуждать логически. В каких случаях на ленте может мигать первый диод а остальное оставаться статичным?
таких случаев два:
1) в ленту отправляется корректно полный буфер в котором меняются только данные для последнего леда
2) в ленту отправляется какой-то мусор с некорректными таймингами, первый чип сбивается с синхронизации и не пропускает данные дальше

первая проблема лежит на стороне пользовательского кода, вторая на сторе фастлед или вернее поведения ядра, которе не может обеспечить требуемую синхроность ногодрыга. Выяснить какой из случаем имеет место быть довольно просто при наличии логического анализатора, но он не у всех есть. Но не суть на самом деле.
Есть одно важное "Но". Во втором случае диод будет мигать всегда и на простых примерах и на эффектах или всегда при каком-то условии, например работающий вайфай.
В случае когда один эффект работает нормально, а при переключении на другой тут же превращается в мигающий светодиод обвинять ядро довольно сложно. Ни ядро ни фастлед не знает что вы там у себя переключили эффект, оно просто продолжает гнать данные из буфера который, который вы же и указали.


> Можете найти ошибку в моем коде, приводящую к указанному эффекту - найдите и ткните носом, тем более, что у себя вы ее воспроизводили, знаете что и где может пойти не так

Не думаю что получится поправить где-то одну строку и проблема уйдет. Скорее придется перерабатывать подходы в работе и переключении эффектов. Вы не очень любите рефакторить текущий код, а какие-то большие изменения вообще вгоняют вас в тоску и отрицалово. Если не готовы вкладывать в это время, то не стоит и начинать, проще мириться с мелкими багами если они не очень часто докучают.

**Reply by vvip-68** [reply, marked as answer]
По ленте:
1. Не последний, а первый диод. Причем, формирование изображения меняется от угла подключения и из угла.  Странно было бы, чтобы картинка поменялась где-то в середине картинки (один диод), а весь остальной буфер изменения картинку не затронули). Тогда в зависмости от угла подключения мигал бы единственный диод где-то в середине матрицы.
2. Тайминги не зависят от содержимого RGB цветов в байтах. Любой мусор просто бы отображался хаотичным набором цветов.

Нет, перерабатывать весь код я не буду, оставим как есть. Это довольно бессмысленное занятие - делать что-нибудь, не понимая что именно. Я могу потратить год времени и перелопатить весь код, при этом оставляя ту же проблему, поскольку я не понимаю в чем она. А значит - год работы с нулевым результатом. 

Я и сейчас уверен, что никаких проблем с выделением / освобождением памяти и неверными указателями (не в тот кусок памяти формируется картинка) в коде скетча нет - просматривал на 10 раз.  Если бы была ошибка формирования картинки "не туда" - она бы проявлялась стабильно. Запустил эффект - картинка пропала. А поскольку в значительном большинстве перекомпиляций картинка все же выводится - значит не в этом проблема.

Программ без багов не бывает. И это относится как к моему скетчу, так и к библиотекам, в том числе - FastLED. Посмотреть хотя бы сколько Issues заявлено в их репозитории.

Дальнейшее продолжение обсуждения, по видимому, бессмысленны.  Вы не можете сказать где точно проблема, чтобы ее исправить. Я тоже не вижу мест в которых может пойти что-то не так. Видел бы - исправил бы. А значит - это всё пустое сотрясание воздуха.

**Reply by vortigont** [reply]
> 1. Не последний, а первый диод

я не говорил "последний", назовем его "Диодом который подключен входом к контроллеру".

> 2\. Тайминги не зависят от содержимого RGB цветов в байтах

нет конечно, но от этого зависит что пропускает через себя дальше "диод который подключен входом к контроллеру".

Но без данных логического анализатора воздух сотрясать бессмысленно, вы правы. У меня есть анализатор, но подобная проблема с моргающим диодом у меня не воспроизводится.
На этом и остановимся.
